File: app/services/insights/adapters/postgres.rb
module Insights::Adapters
  class Postgres < SQL
    def convert_sql_timezone(sql)
      sql = "(#{sql} at time zone 'UTC' at time zone '#{Time.zone.name}')"
    end

    def allowed_date_truncations
      %w(hour day week month quarter year)
    end

    def truncate_date(sql, truncation)
      raise "Bad date truncation '#{truncation}'" unless truncation.to_s.in(allowed_date_truncations)
      "date_trunc('#{date_part}', #{sql})::date"
    end

    def filter_contains(sql, string)
      "(#{sql}) ilike #{quote('%' + string + '%')}"
    end
  end
end

File: old-packages/insights-core-old/app/insights/adapter/sql/postgres.js
const { Pool } = require('pg')
const escape = require('pg-escape')

const SQL = require('./index')

const { timezone } = require('../../../../config/insights')

module.exports = class Postgres extends SQL {
  constructor (connection) {
    super(connection)

    this.pool = new Pool({
      connectionString: connection
    })
  }

  quote (string) {
    return escape('%L', string)
  }

  async execute (sql) {
    try {
      console.log('Executing', sql)
      return await this.pool.query(sql)
    } catch (error) {
      console.error('Error with SQL', sql)
      console.error(error)
      throw error
    }
  }

  convertSqlTimezone (sql) {
    return `(${sql} at time zone 'UTC' at time zone '${timezone}')`
  }

  allowedDateTruncations () {
    return ['hour', 'day', 'week', 'month', 'quarter', 'year']
  }

  truncateDate (sql, truncation) {
    if (!this.allowedDateTruncations().includes(truncation)) {
      throw new Error(`Bad date truncation '${truncation}'`)
    }

    return `date_trunc('${truncation}', ${sql})::date`
  }

  filterContains (sql, string) {
    return `(${sql}) ilike ${this.quote('%' + string + '%')}`
  }
}
